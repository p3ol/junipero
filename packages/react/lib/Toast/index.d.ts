import { ElementType, MutableRefObject } from 'react';

export declare type ToastRef = {
  innerRef: MutableRefObject<any>;
  enabled: boolean;
  paused: boolean;
  remaining: number;
  isJunipero: boolean;
};

export declare interface ToastObject {
  index?: string|number;
  content?: string | React.ReactNode | Function;
  duration?: number;
  lifespan?: number;
  onDismiss?: (index?: string|number) => any;
}

declare interface ToastProps extends React.ComponentPropsWithRef<any> {
  animate?: (
    alert: JSX.Element,
    opts: { opened: boolean, index: string|number }
  ) => JSX.Element;
  animationTimeout?: number;
  className?: string;
  index?: string|number;
  lifespan?: number;
  tag?: String | ElementType;
  children?: string | React.ReactNode | Function;
  onDismiss?(index?: string|number): any;
  ref?: MutableRefObject<ToastRef | undefined>;
}

declare function Toast(props: ToastProps): JSX.Element;

export default Toast;
