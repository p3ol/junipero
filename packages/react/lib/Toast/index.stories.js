import Toast from '.';

export default { title: 'react/Toast' };

export const basic = () => (
  <Toast lifespan={5000}>Toast</Toast>
);
