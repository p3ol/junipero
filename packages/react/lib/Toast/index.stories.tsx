import Toast from '.';

export default { title: 'react/Toast' };

export const Basic = () => (
  <Toast lifespan={5000}>Toast</Toast>
);
