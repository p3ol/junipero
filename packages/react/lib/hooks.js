import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
