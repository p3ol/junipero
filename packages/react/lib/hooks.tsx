import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
