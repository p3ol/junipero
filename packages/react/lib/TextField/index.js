import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState, exists } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';

const TextField = forwardRef(({
  min,
  max,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val, { required }) => !!val || !required,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = value => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = e => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = e => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = e => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = dirty => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        min={min}
        max={max}
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? (() => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
});

TextField.displayName = 'TextField';
TextField.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  type: PropTypes.string,
  valid: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onWheel: PropTypes.func,
  onValidate: PropTypes.func,
};

export default TextField;
