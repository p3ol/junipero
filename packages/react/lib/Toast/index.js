import { forwardRef, useState, useImperativeHandle, useRef } from 'react';
import { classNames } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';
import PropTypes from 'prop-types';

const Toast = forwardRef(({
  index,
  tag: Tag = 'div',
  animationTimeout = 100,
  pausable = true,
  lifespan = 0,
  animate,
  children,
  className,
  onDismiss,
  onClick,
  onMouseEnter,
  onMouseLeave,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const startTimeRef = useRef(Date.now() + animationTimeout);
  const [remaining, setRemaining] = useState(lifespan);
  const [enabled, setEnabled] = useState(true);
  const [paused, setPaused] = useState(false);

  useImperativeHandle(ref, () => ({
    enabled,
    paused,
    remaining,
    innerRef,
    isJunipero: true,
  }));

  useTimeout(() => {
    setEnabled(false);
  }, remaining, [remaining, paused], { enabled: lifespan > 0 && !paused });

  useTimeout(() => {
    onDismiss?.(index);
  }, animationTimeout, [enabled], { enabled: !enabled && !paused });

  const onClick_ = e => {
    onClick?.(e);
    setEnabled(false);
  };

  const onMouseEnter_ = e => {
    onMouseEnter?.(e);

    if (!pausable) {
      return;
    }

    setPaused(true);
    setRemaining(remaining - (Date.now() - startTimeRef.current));
  };

  const onMouseLeave_ = e => {
    onMouseLeave?.(e);

    if (!pausable) {
      return;
    }

    startTimeRef.current = Date.now();
    setPaused(false);
  };

  const content = (
    <Tag
      index={index}
      ref={innerRef}
      className={classNames('junipero toast', className)}
      onClick={onClick_}
      onMouseEnter={onMouseEnter_}
      onMouseLeave={onMouseLeave_}
      { ...rest }
    >
      { children }
      { lifespan > 0 && (
        <div
          className="countdown"
          style={{
            animationDuration: `${lifespan}ms`,
            animationPlayState: paused ? 'paused' : 'running',
          }}
        />
      ) }
    </Tag>
  );

  return animate ? animate(content, { opened: enabled, index }) : content;
});

Toast.displayName = 'Toast';
Toast.propTypes = {
  index: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  lifespan: PropTypes.number,
  animate: PropTypes.func,
  animationTimeout: PropTypes.number,
  tag: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
    PropTypes.node,
  ]),
  pausable: PropTypes.bool,
  onDismiss: PropTypes.func,
  onClick: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
};

export default Toast;
