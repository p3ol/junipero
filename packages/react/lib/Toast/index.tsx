import {
  type ReactNode,
  type ElementType,
  type MouseEvent,
  useState,
  useImperativeHandle,
  useRef,
} from 'react';
import { classNames } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type {
  JuniperoRef,
  SpecialComponentPropsWithRef,
} from '../types';
import type { TransitionProps } from '../Transition';

export declare interface ToastRef extends JuniperoRef {
  enabled: boolean;
  paused: boolean;
  remaining: number;
}

export declare interface ToastObject {
  animationTimeout?: number;
  content?: ReactNode;
  duration?: number;
  index?: string | number;
  lifespan?: number;
  animate?(
    alert: ReactNode,
    opts: {
      opened: boolean;
      index: string | number;
    } & Partial<TransitionProps>,
  ): ReactNode;
  onDismiss?(index?: string | number): any;
}

export declare interface ToastProps
  extends SpecialComponentPropsWithRef<'div', ToastRef> {
  animationTimeout?: number;
  index?: string | number;
  lifespan?: number;
  pausable?: boolean;
  tag?: string | ElementType;
  animate?(
    alert: ReactNode,
    opts?: {
      opened: boolean;
      index: string | number;
    } & Partial<TransitionProps>,
  ): ReactNode;
  onDismiss?(index?: string | number): any;
}

const Toast = ({
  ref,
  index,
  children,
  className,
  tag: Tag = 'div',
  animationTimeout = 100,
  pausable = true,
  lifespan = 0,
  animate,
  onClick,
  onDismiss,
  onMouseEnter,
  onMouseLeave,
  ...rest
}: ToastProps) => {
  const timeout = animate ? animationTimeout : 0;
  const innerRef = useRef<HTMLElement>(null);
  const startTimeRef = useRef(Date.now() + timeout);
  const [remaining, setRemaining] = useState(lifespan);
  const [enabled, setEnabled] = useState(true);
  const [paused, setPaused] = useState(false);

  useImperativeHandle(ref, () => ({
    enabled,
    paused,
    remaining,
    innerRef,
    isJunipero: true,
  }));

  useTimeout(() => {
    setEnabled(false);
  }, remaining, [remaining, paused], { enabled: lifespan > 0 && !paused });

  useTimeout(() => {
    onDismiss?.(index);
  }, timeout, [enabled], { enabled: !enabled && !paused });

  const onClick_ = (e: MouseEvent<HTMLDivElement>) => {
    onClick?.(e);
    setEnabled(false);

    if (!animate) {
      onDismiss?.(index);
    }
  };

  const onMouseEnter_ = (e: MouseEvent<HTMLDivElement>) => {
    onMouseEnter?.(e);

    if (!pausable) {
      return;
    }

    setPaused(true);
    setRemaining(remaining - (Date.now() - startTimeRef.current));
  };

  const onMouseLeave_ = (e: MouseEvent<HTMLDivElement>) => {
    onMouseLeave?.(e);

    if (!pausable) {
      return;
    }

    startTimeRef.current = Date.now();
    setPaused(false);
  };

  const content = (
    <Tag
      ref={innerRef}
      className={classNames('junipero toast', className)}
      onClick={onClick_}
      onMouseEnter={onMouseEnter_}
      onMouseLeave={onMouseLeave_}
      { ...rest }
    >
      { children }
      { lifespan > 0 && (
        <div
          className="countdown"
          style={{
            animationDuration: `${lifespan}ms`,
            animationPlayState: paused ? 'paused' : 'running',
          }}
        />
      ) }
    </Tag>
  );

  return animate ? animate(content, { opened: enabled, index })
    : content;
};

Toast.displayName = 'Toast';

export default Toast;
