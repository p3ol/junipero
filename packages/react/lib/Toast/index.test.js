import { render, act, fireEvent } from '@testing-library/react';

import ToastsControl from '../ToastsControl';
import Toasts from '../Toasts';
import Toast from './index';

describe('<Toast />', () => {
  it('should render', () => {
    const { container, unmount } = render(<Toast>Content</Toast>);
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should be auto dismissable', async () => {
    jest.useFakeTimers();
    const toast = { id: 1, content: 'Content', lifespan: 100 };
    const { container, unmount } = render(
      <ToastsControl toasts={[toast]}>
        <Toasts />
      </ToastsControl>
    );
    expect(container).toMatchSnapshot();

    await act(async () => jest.advanceTimersByTime(100));
    await act(async () => jest.advanceTimersByTime(1));
    expect(container).toMatchSnapshot();

    unmount();
  });

  it('should be manually dismissable', async () => {
    const toast = { id: 1, content: 'Content', lifespan: 100 };
    const { container, unmount } = render(
      <ToastsControl toasts={[toast]}>
        <Toasts />
      </ToastsControl>
    );
    expect(container).toMatchSnapshot();

    fireEvent.click(container.querySelector('.toast'));

    expect(container).toMatchSnapshot();

    unmount();
  });
});
