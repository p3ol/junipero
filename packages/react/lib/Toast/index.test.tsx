import { render, act, fireEvent } from '@testing-library/react';

import ToastsControl from '../ToastsControl';
import Toasts from '../Toasts';
import Toast, { type ToastObject } from './index';

describe('<Toast />', () => {
  it('should render', () => {
    const { container, unmount } = render(<Toast>Content</Toast>);
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should be auto dismissable', async () => {
    jest.useFakeTimers();
    const toast: ToastObject = { index: 1, content: 'Content', lifespan: 100 };
    const { container, unmount } = render(
      <ToastsControl toasts={[toast]}>
        <Toasts />
      </ToastsControl>
    );
    expect(container).toMatchSnapshot();

    await act(async () => Promise.resolve(jest.advanceTimersByTime(100)));
    await act(async () => Promise.resolve(jest.advanceTimersByTime(1)));
    expect(container).toMatchSnapshot();

    unmount();
  });

  it('should be manually dismissable', () => {
    const toast: ToastObject = { index: 1, content: 'Content', lifespan: 100 };
    const { container, unmount } = render(
      <ToastsControl toasts={[toast]}>
        <Toasts />
      </ToastsControl>
    );
    expect(container).toMatchSnapshot();

    fireEvent.click(container.querySelector('.toast'));

    expect(container).toMatchSnapshot();

    unmount();
  });
});
