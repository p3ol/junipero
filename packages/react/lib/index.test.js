import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
