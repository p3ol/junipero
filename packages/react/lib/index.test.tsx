import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
