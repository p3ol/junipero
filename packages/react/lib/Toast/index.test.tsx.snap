// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<Toast /> should be auto dismissable 1`] = `
<div>
  <div
    class="junipero toasts"
  >
    <div
      class="junipero toast"
    >
      Content
      <div
        class="countdown"
        style="animation-duration: 100ms; animation-play-state: running;"
      />
    </div>
  </div>
</div>
`;

exports[`<Toast /> should be auto dismissable 2`] = `
<div>
  <div
    class="junipero toasts"
  />
</div>
`;

exports[`<Toast /> should be manually dismissable 1`] = `
<div>
  <div
    class="junipero toasts"
  >
    <div
      class="junipero toast"
    >
      Content
      <div
        class="countdown"
        style="animation-duration: 100ms; animation-play-state: running;"
      />
    </div>
  </div>
</div>
`;

exports[`<Toast /> should be manually dismissable 2`] = `
<div>
  <div
    class="junipero toasts"
  />
</div>
`;

exports[`<Toast /> should render 1`] = `
<div>
  <div
    class="junipero toast"
  >
    Content
  </div>
</div>
`;
